<launch>

  <!-- <include file="$(find velodyne_pointcloud)/launch/VLP16_points.launch"/> -->

  <!-- <node pkg="rviz" type="rviz" name="rviz" args=""/>  -->

  <!-- <node pkg="Lidar_MCduo_2023" type="Lidar_Downsampling" name="Downsampling_node" respawn="true" output="screen">
    <param name="voxel_size_x" value="2"/>
    <param name="voxel_size_y" value="2"/>
    <param name="voxel_size_z" value="2"/>
  </node> -->

  <node pkg="Lidar_MCduo_2023" type="Lidar_UDP" name="UDP_node" respawn="true" output="screen">
    <param name="UDP_IP" value="127.0.0.1"/>
    <param name="Port_Num" value="2368"/>
    <param name="GE_Z" value="-0.6"/>
    <param name="GE_Z_4" value="-0.5"/>
    <param name="GE_Z_re" value="-0.6"/>
    <param name="GE_slope" value="0.3"/>
  </node>
  <!-- 192.168.1.85 -->                                                                  

  <node pkg="Lidar_MCduo_2023" type="Lidar_ROI" name="ROI_node" respawn="true" output="screen">
  <!--패키지 명, type => excutable file, cpp파일에 ROI_node로 되어있음-->
    <!--ROI parameter-->
    <param name="switch_ROI" value="true"/>
    <param name="ROI_xMin" value="0.5"/>
    <param name="ROI_xMax" value="15"/>
    <param name="ROI_yMin" value="-10"/>
    <param name="ROI_yMax" value="10"/>
    <param name="ROI_zMin" value="-2"/> <!-- ransac auto off (when up to ROI_zMin > -0.2) -->
    <param name="ROI_zMax" value="5"/> <!-- 5-->
    <!-- signal -->
    <param name="ROI_zMin_2" value="0.17"/> <!-- 표지판이 지면으로부터 1m (라이다 83cm에 위치)--> 
    <param name="ROI_yMin_2" value="-10"/>
    <param name="ROI_yMax_2" value="10"/>

    <param name="ROI_xMax_4" value="12"/>
    <param name="ROI_yMin_4" value="-15"/>
    <param name="ROI_yMax_4" value="15"/>

    <param name="ROI_xMin_8" value="-0.5"/>
    <param name="ROI_xMax_8" value="12"/> <!-- 12 -->
    <param name="ROI_yMin_8" value="-5.5"/> <!-- -5.5 -->
    <param name="ROI_yMax_8" value="5.5"/> <!-- 5.5 -->
    <param name="ROI_zMax_8" value="0.8"/> <!-- 0.8 -->

    <param name="ROI_xMin_re" value="0.5"/>
    <param name="ROI_xMax_re" value="15"/>
    <param name="ROI_yMin_re" value="-5.5"/>
    <param name="ROI_yMax_re" value="5.5"/>
    <param name="ROI_zMin_re" value="-2"/>
    <param name="ROI_zMax_re" value="5"/>
  </node>

  <node pkg="Lidar_MCduo_2023" type="Lidar_ransac" name="ransac_node" respawn="true" output="screen">
    <!--ransac parameter-->
    <param name="switch_UpSampling" value="false"/>
    <param name="switch_RanSaC" value="false"/>
    <param name="ransac_distanceThreshold" value="0.85"/> <!--1-->
  </node>

  <node pkg="Lidar_MCduo_2023" type="Lidar_Clustering" name="Clustering_node" respawn="true" output="screen">
	
    <!--EuclidClustering parameter-->
    <param name="switch_Euclid" value="false"/>
    <param name="EC_eps" value="0.6"/>
    <param name="EC_MinClusterSize" value="5"/>
    <param name="EC_MaxClusterSize" value="100000"/>

    <!--DBSCANclustering parameter-->
    <param name="switch_DBscan" value="true"/>
    <param name="DBscan_eps" value="0.5"/>
    <param name="DBscan_minPts" value="3"/>
    <param name="DB_MinClusterSize" value="5"/>
    <param name="DB_MaxClusterSize" value="10000"/>

    <!-- signal -->
    <param name="DBscan_eps_1" value="0.5"/>
    <param name="DBscan_minPts_1" value="3"/>
    <param name="DB_MinClusterSize_1" value="5"/>
    <param name="DBscan_eps_2" value="0.35"/>
    <param name="DBscan_minPts_2" value="3"/>
    <param name="DB_MinClusterSize_2" value="4"/>
    <param name="DBscan_eps_3" value="0.5"/>
    <param name="DBscan_minPts_3" value="3"/>
    <param name="DB_MinClusterSize_3" value="5"/>
    <param name="DBscan_eps_4" value="0.5"/>
    <param name="DBscan_minPts_4" value="3"/>
    <param name="DB_MinClusterSize_4" value="5"/>  <!-- 5 -->
    <param name="DBscan_eps_8" value="0.35"/>
    <param name="DBscan_minPts_8" value="3"/>      <!-- 3 -->
    <param name="DB_MinClusterSize_8" value="3"/>  <!-- 3 -->

    <param name="DBscan_eps_re" value="0.5"/>
    <param name="DBscan_minPts_re" value="3"/>
    <param name="DB_MinClusterSize_re" value="5"/>

    <param name="REMOVE_FACTOR_1" value="1.5"/>
    <param name="REMOVE_FACTOR_3" value="1"/>
    <param name="REMOVE_FACTOR_4" value="1"/>
    <param name="REMOVE_FACTOR_8" value="0.35"/> <!-- 라바콘 한개는 그대로 잡되, 잔디같은 뭉칠 수 있는애들은 TC필터로 거르기 위함. 0.4가 라바콘 두개 딱 붙였을때 수치임. -->
    <param name="REMOVE_FACTOR_re" value="1.5"/>

    <!--NoiseFiltering parameter-->
    <param name="switch_NoiseFiltering" value="false"/>

    <!--Filter on/off-->    
    <param name="switch_jiwon_filter" value="true"/>
    <param name="switch_DY_filter" value="false"/>
    <param name="switch_TC_filter"      value="false"/>

    <!--jiwonFilter parameter-->
    <param name="REMOVE_FACTOR" value="1.5"/>

    <!--DYFilter parameter-->
    <param name="Ransac_Z_ROI" value="-0.55"/>
  </node>

  <node pkg="Lidar_MCduo_2023" type="Lidar_visual_2D" name="visual_2D_node" output="screen">
    <param name="switch_visual_2D" value="true"/>
    <param name="location" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/plane.png"/>
  </node>

  <!-- <node pkg="Lidar_MCduo_2023" type="Lidar_visual_map" name="visual_map_node" respawn="true" output="screen">
    <param name="switch_visual_map" value="true"/>
    <param name="path_location" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/kcity Main Path.txt"/>
    <param name="path_location1" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/prl_1.txt"/>
    <param name="path_location2" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/prl_2.txt"/>
    <param name="path_location3" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/prl_3.txt"/>
    <param name="path_location4" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/LBC.txt"/> -->
    <!--kcity Main Path.txt / -->
    <!-- <param name="image_location" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/K_city_main(height 597.4)(downsized 2).png"/> -->
    <!--K_city_main(height 597.4)(downsized 2).png / K_city_qual(height 316).png / -->
  <!-- </node> -->

  <node pkg="Lidar_MCduo_2023" type="Lidar_TS_filter" name="TS_filter_node" respawn="true" output="screen"></node>

  <!-- <node pkg="Lidar_MCduo_2023" type="Lidar_tracking" name="tracking_node" output="screen"></node> -->

  <node pkg="Lidar_MCduo_2023" type="Lidar_tetragon_interior_points" name="tetragon_interior_points_node" respawn="true" output="screen"></node>

  <node pkg="Lidar_MCduo_2023" type="Lidar_tracking_global" name="tracking_global_node" respawn="true" output="screen">
    <param name="P_0_0" value="2"/>
    <param name="P_1_1" value="2"/>
    <param name="P_2_2" value="3"/>
    <param name="P_3_3" value="3"/>
    <param name="Q_0_0" value="0.5"/>
    <param name="Q_1_1" value="0.5"/>
    <param name="Q_2_2" value="2"/>
    <param name="Q_3_3" value="2"/>
    <param name="R_0_0" value="0.01"/>
    <param name="R_1_1" value="0.01"/>
    <param name="R_2_2" value="1.5"/>
    <param name="R_3_3" value="1.5"/>

    <param name="max_erase_cnt" value="30"/>
    <param name="min_near_dist" value="2"/>
    <param name="min_new_dist" value="1"/>
    <param name="max_pred_dist" value="6"/>
    <param name="max_pred_vel" value="150"/>

    <param name="where_are_you" value="$(find Lidar_MCduo_2023)/include/Lidar_MCduo_2023/Map_data/kcity Reference LLA.txt"/>
  </node>
  <!--K_city_main / K_city_qual / SNU_Siheung-->
</launch>
